streams/src/main/java/org/apache/kafka/streams/state/internals/MeteredTimestampedKeyValueStore.java

```diff
@@ -102,4 +139,232 @@ static class RawAndDeserializedValue<ValueType> {
+        final QueryResult<R> result;
+
+        final StoreQueryUtils.QueryHandler handler = queryHandlers.get(query.getClass());
+        if (handler == null) {
+            result = wrapped().query(query, positionBound, config);
+            if (config.isCollectExecutionInfo()) {
+                result.addExecutionInfo(
+                        "Handled in " + getClass() + " in " + (time.nanoseconds() - start) + "ns");
+            }
+        } else {
+            result = (QueryResult<R>) handler.apply(
+                    query,
+                    positionBound,
+                    config,
+                    this
+            );
+            if (config.isCollectExecutionInfo()) {
+                result.addExecutionInfo(
+                        "Handled in " + getClass() + " with serdes "
+                                + serdes + " in " + (time.nanoseconds() - start) + "ns");
+            }
+        }
+        return result;
+    }
+
+
+
+    @SuppressWarnings("unchecked")
+    protected <R> QueryResult<R> runTimestampedKeyQuery(final Query<R> query,
+                                                      final PositionBound positionBound,
+                                                      final QueryConfig config) {
+        final QueryResult<R> result;
+        final TimestampedKeyQuery<K, V> typedKeyQuery = (TimestampedKeyQuery<K, V>) query;
+        final KeyQuery<Bytes, byte[]> rawKeyQuery =
+                KeyQuery.withKey(keyBytes(typedKeyQuery.key()));
+        final QueryResult<byte[]> rawResult =
+                wrapped().query(rawKeyQuery, positionBound, config);
+        if (rawResult.isSuccess()) {
+            final Function<byte[], ValueAndTimestamp<V>> deserializer = getDeserializeValue(serdes, wrapped());
+            final ValueAndTimestamp<V> value = deserializer.apply(rawResult.getResult());
+            final QueryResult<ValueAndTimestamp<V>> typedQueryResult =
+                    InternalQueryResultUtil.copyAndSubstituteDeserializedResult(rawResult, value);
+            result = (QueryResult<R>) typedQueryResult;
+        } else {
+            // the generic type doesn't matter, since failed queries have no result set.
+            result = (QueryResult<R>) rawResult;
+        }
+        return result;
+    }
+
+    @SuppressWarnings("unchecked")
+    protected <R> QueryResult<R> runTimestampedRangeQuery(final Query<R> query,
+                                                        final PositionBound positionBound,
+                                                        final QueryConfig config) {
+
+        final QueryResult<R> result;
+        final TimestampedRangeQuery<K, V> typedQuery = (TimestampedRangeQuery<K, V>) query;
+        RangeQuery<Bytes, byte[]> rawRangeQuery;
+        final boolean isKeyAscending = typedQuery.isKeyAscending();
+        if (typedQuery.lowerBound().isPresent() && typedQuery.upperBound().isPresent()) {
+            rawRangeQuery = RangeQuery.withRange(
```

Comment:
Given that we can pass in `null` for upper and/or lower bound (with semantics "no bound") I am wondering if we can simplify this whole if/else stuff here?
```
rawRangeQuery = RangeQuery.withRange(
    keyBytes(typedQuery.lowerBound().orElse(null)),
    keyBytes(typedQuery.upperBound().orElse(null))
);
```

Reply:
Might be a good cleanup for `runRangeQuery(...)` too (in here, as well as `MeteredKeyValueStore`) -- when `RangeQuery` was added, passing in `null` for upper/lower was not supported to the code needed to be written in this complex way.

Reply:
Yes, I reuse the code in the` MeteredKeyValueStore`.